src/morse/ColorMorseReader.java

```diff
@@ -1,41 +1,144 @@
 package morse;
 
-import lejos.hardware.lcd.LCD;
+import hardware.MotorController;
 import lejos.hardware.sensor.EV3ColorSensor;
-import lejos.utility.Delay;
+import lejos.hardware.sensor.SensorMode;
+import hardware.ConfiguredTimer;
+import lejos.hardware.ev3.LocalEV3;
+import lejos.hardware.lcd.TextLCD;
+import lejos.hardware.port.Port;
+import lejos.hardware.port.SensorPort;
+import lejos.hardware.port.UARTPort;
 
 public class ColorMorseReader extends Thread implements MorseReader {
+    private static final float DOT_MAX = 600;
+    private static final float DASH_MIN = 900;
+    private static final float WORD_SPACE = 8000;
+    private static final float TERMINATION_THRESHOLD = 20000;
+    private static final float BLACK_THRESH = 0.3f;
+    private static final float LETTER_SPACE = 3000;
 
-    float DOT_DUR;
-    float DASH_DUR;
-    float SPACE_DUR;
-    private EV3ColorSensor COLOR_SEN;
-    private float[] CURR_SAMPLE;
-    public StringBuilder WORD;
-    public StringBuilder MORSE;
+    private StringBuilder MORSE = new StringBuilder();
+    private StringBuilder WORD = new StringBuilder();
+    private MotorController motorController;
+    private final ConfiguredTimer blackTimer;
+    private final SensorMode colorMode;
+    private long lastBlackTime;
     private boolean running;
-    private float BLACK_MIN;
-    private float BLACK_MAX;
+    private final EV3ColorSensor colorSensor;
+    private StringBuilder MorseSequence = new StringBuilder(); // Used to store the full Morse sequence
+
+    public ColorMorseReader(MotorController motorController) {
+        this.motorController = motorController;
+        colorSensor = new EV3ColorSensor(SensorPort.S2);
```

Comment:
Initiate ColSen in the Morsey main and then pass it to the Reader
```suggestion
        colorSensor = new EV3ColorSensor(SensorPort.S2);
```